src/Npgsql/NpgsqlSchema.cs

```diff
@@ -563,6 +566,248 @@ static DataTable GetDataSourceInformation(NpgsqlConnection conn)
+            table.Columns.Add("IsAutoIncrementable", typeof(bool));
+            table.Columns.Add("IsBestMatch", typeof(bool));
+            table.Columns.Add("IsCaseSensitive", typeof(bool));
+            table.Columns.Add("IsConcurrencyType", typeof(bool));
+            table.Columns.Add("IsFixedLength", typeof(bool));
+            table.Columns.Add("IsFixedPrecisionAndScale", typeof(bool));
+            table.Columns.Add("IsLiteralSupported", typeof(bool));
+            table.Columns.Add("IsLong", typeof(bool));
+            table.Columns.Add("IsNullable", typeof(bool));
+            table.Columns.Add("IsSearchable", typeof(bool));
+            table.Columns.Add("IsSearchableWithLike", typeof(bool));
+            table.Columns.Add("IsUnsigned", typeof(bool));
+            table.Columns.Add("LiteralPrefix", typeof(string));
+            table.Columns.Add("LiteralSuffix", typeof(string));
+            table.Columns.Add("MaximumScale", typeof(short));
+            table.Columns.Add("MinimumScale", typeof(short));
+            table.Columns.Add("NativeDataType", typeof(string));
+            table.Columns.Add("ProviderDbType", typeof(int));
+
+            // Npgsql-specific
+            table.Columns.Add("OID", typeof(uint));
+
+            // TODO: Support type name restriction
+
+            foreach (var baseType in connector.DatabaseInfo.BaseTypes)
+            {
+                if (!connector.TypeMapper.Mappings.TryGetValue(baseType.Name, out var mapping) &&
+                    !connector.TypeMapper.Mappings.TryGetValue(baseType.FullName, out mapping))
+                    continue;
+
+                var row = table.Rows.Add();
+
+                PopulateDefaultDataTypeInfo(row, baseType);
+                PopulateHardcodedDataTypeInfo(row, baseType);
+
+                if (mapping.ClrTypes.Length > 0)
+                    row["DataType"] = mapping.ClrTypes[0].Name;
+                if (mapping.NpgsqlDbType.HasValue)
+                    row["ProviderDbType"] = (int)mapping.NpgsqlDbType.Value;
+            }
+
+            foreach (var arrayType in connector.DatabaseInfo.ArrayTypes)
+            {
+                if (!connector.TypeMapper.Mappings.TryGetValue(arrayType.Element.Name, out var elementMapping) &&
+                    !connector.TypeMapper.Mappings.TryGetValue(arrayType.Element.FullName, out elementMapping))
+                    continue;
+
+                var row = table.Rows.Add();
+
+                PopulateDefaultDataTypeInfo(row, arrayType.Element);
+                // Populate hardcoded values based on the element type (e.g. citext[] is case-insensitive).
+                PopulateHardcodedDataTypeInfo(row, arrayType.Element);
+
+                row["TypeName"] = arrayType.DisplayName;
+                row["OID"] = arrayType.OID;
+                row["CreateFormat"] += "[]";
+                if (elementMapping.ClrTypes.Length > 0)
+                    row["DataType"] = elementMapping.ClrTypes[0].MakeArrayType();
```

Comment:
You're right. I suspect this works because DataTable probably does ToString() on arbitrary types it does not recognize... Will fix.

Reply:
I've switched to using `Type.FullName`. Unfortunately for `NpgsqlRange<T>` this produces an assembly-qualified name for the type parameter (including the version and signature :/), but for 100% correctness it's probably the right way to go - you should be able to "dynamically" load the string you get and it will probably always work.

Let me know what you think.